File: Graphs/785. Is Graph Bipartite.py
import collections
class Solution:
    # dfs 
    def isBipartite(self, graph):
        colorings = {}
        def dfs(currentNode):
            for i in graph[currentNode]:
                if i in colorings:
                    if colorings[i] == colorings[currentNode]:
                        return False 
                else:
                    colorings[i] = colorings[currentNode] ^ 1 
                    if not dfs(i):
                        return False 
                    
            return True 
        
        for idx in range(len(graph)):
            if idx not in colorings:
                
                colorings[idx] = 1 
                if not dfs(idx):
                    return False
                
        return True

File: Graphs/785. Is Graph Bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> colors(n, 0);

        for (int i = 0; i < n; i++) {
            if (colors[i] == 0) {
                if (!dfs(i, 1, colors,graph)) return false;
            }
        }

        return true;

    }

    bool dfs(int node, int color, vector<int>& colors, vector<vector<int>>& graph) {
        if (colors[node] != 0) {
            // check if the colors match the expected color here

            if (colors[node] != color) return false;

            return true;
        }

        colors[node] = color;
        for (int i = 0; i < graph[node].size(); i++) {
            if (color == 1) {
                if (!dfs(graph[node][i], 2, colors, graph)) return false;
            } else {
                if (!dfs(graph[node][i], 1, colors, graph)) return false;
            }
        }

        return true;


    }
};